src/utils/helpers/structure/generateProgress.ts

```diff
@@ -0,0 +1,82 @@
+import { TemplateElementCategory } from '../../generated/graphql'
+import { ElementStateNEW, FullStructure, PageNEW, Progress } from '../../types'
+
+const initialProgress = {
+  doneNonRequired: 0,
+  doneRequired: 0,
+  completed: false,
+  totalNonRequired: 0,
+  totalRequired: 0,
+  totalSum: 0,
+  valid: true,
+}
+
+const calculateCompleted = (totalSum: number, doneRequired: number, doneNonRequired: number) => {
+  const totalDone = doneRequired + doneNonRequired
+  return totalSum === totalDone
+}
+
+const getSectionProgress = (pages: PageNEW[]): Progress => {
+  return pages.reduce((sectionProgress: Progress, { progress }) => {
+    if (!progress) return sectionProgress
+    let {
+      doneNonRequired,
+      doneRequired,
+      completed,
+      totalNonRequired,
+      totalRequired,
+      totalSum,
+      valid,
+    } = sectionProgress
+    doneNonRequired += progress.doneNonRequired || 0
+    totalNonRequired += progress.totalNonRequired || 0
+    doneRequired += progress.doneRequired || 0
+    totalRequired += progress.totalRequired || 0
+    totalSum = totalRequired + doneNonRequired
+    completed = calculateCompleted(totalSum, doneRequired, doneNonRequired)
+    if (!progress.valid) valid = false
+    return {
+      doneNonRequired,
+      doneRequired,
+      completed,
+      totalNonRequired,
+      totalRequired,
+      totalSum,
+      valid,
+    }
+  }, initialProgress)
+}
+
+export const generateResponsesProgress = (structure: FullStructure) => {
+  Object.values(structure.sections).forEach((section) => {
+    Object.values(section.pages).forEach((page) => {
+      const progress: Progress = initialProgress
+      page.state
+        .filter(({ element }) => {
+          const { category, isVisible, isEditable } = element as ElementStateNEW
+          return isVisible && isEditable && category === TemplateElementCategory.Question
+        })
+        .forEach(({ element, response }) => {
+          const { isRequired } = element as ElementStateNEW
+          if (isRequired) progress.totalRequired++
+          else progress.totalNonRequired++
+          if (response?.text) {
+            if (isRequired) progress.doneRequired++
+            else progress.doneNonRequired++
+          }
+          if (response !== null && !response?.isValid) progress.valid = false
```

Comment:
Sorry I can't quite remember, i thought that progress.valid = false would only be set where validation caused an actual 'false' value ? (ie if response.isValid = null, it will still be valid ?). I think just doing `typeof response !== 'undefined' && response?.isValid == false` would work ?

![Screen Shot 2021-02-24 at 5 16 40 PM](https://user-images.githubusercontent.com/26194949/108946856-2583e300-76c4-11eb-9389-b34a70f0bb1b.png)

Reply:
Yep. It should be only for actual `isValid` = `false`. Will update to use `typeof response !== 'undefined' && response?.isValid == false` now. Thanks : )